scripts/merge_source_cat.py

```diff
@@ -0,0 +1,415 @@
+            # Restrict ref_table to isPrimary?
+
+            # We get back and array of matching object IDs
+            # Cat rows with no matches get a -1 entry.
+            these_associated_ids = associate_object_ids_to_table(cat, object_table=ref_table,
+                                                                 verbose=verbose, **kwargs)
+            associated_id_table.append(these_associated_ids)
+
+    # Now merge the lists together, taking the last non-negative entry
+    # TODO Could consider checking for an inconsistent entry.
+    associated_ids = np.zeros(len(cat), dtype=int) - 1
+    for matches in associated_id_table:
+        w, = np.where(matches >= 0)
+        associated_ids[w] = matches[w]
+
+    return associated_ids
+
+
+def associate_object_ids_to_table(cat, object_table=None, matching_radius=1,
+                                  verbose=True):
+    """Return object ID associated with each entry in cat.
+
+    Takes closest match within matching radius.
+    Entries with no match will be -1.  (the index entries are integers so we can't use NaN as signaling value)
+
+    Currently does purely position based matching above an implicit SNR
+    Does not do a flux-based matching.
+
+    Parameters
+    --
+    cat: pandas DataFrame
+    object_table:  pandas DataFrame
+    matching_radius: float [arcsec]
+
+    Notes
+    --
+    cat and object_table are assumed to have RA, Dec units of rad
+
+    """
+    # Let's first try the stupid way:
+    # Explicitly iterate through cat
+    # Construct a subset of object
+
+    min_ra, max_ra = np.min(cat['coord_ra']), np.max(cat['coord_ra'])
+    min_dec, max_dec = np.min(cat['coord_dec']), np.max(cat['coord_dec'])
+    edge_buffer = np.deg2rad(10 * matching_radius / 3600)
+    # Note this is wrong.  There's no cos(dec)
+    ra_range = (np.rad2deg(min_ra - edge_buffer), np.rad2deg(max_ra + edge_buffer))
+    dec_range = (np.rad2deg(min_dec - edge_buffer), np.rad2deg(max_dec + edge_buffer))
+
+    if verbose:
+        print("Extract objects from Object Table with RA, Dec in ranges:")
+        print(ra_range, dec_range)
+    in_source_catalog_area = (ra_range[0] < object_table['ra']) & \
+                             (object_table['ra'] < ra_range[1]) & \
+                             (dec_range[0] < object_table['dec']) & \
+                             (object_table['dec'] < dec_range[1])
+    this_object_table = object_table[in_source_catalog_area]
+    if verbose:
+        print("Found %d objects in Object Table in region" % len(this_object_table))
+
+    if len(this_object_table) < 1:
+        # Return an array of -1
+        return np.zeros(len(cat)) - 1
+
+    this_object_table_skyCoor = SkyCoord(this_object_table['ra'],
+                                         this_object_table['dec'],
+                                         unit=(u.deg, u.deg))
+    cat_skyCoor = SkyCoord(cat['coord_ra'], cat['coord_dec'], unit=(u.rad, u.rad))
```

Comment:
Just to clarify, alternatively `SkyCoord(this_object_table['coord_ra'], this_object_table['coord_dec'], unit=(u.rad, u.rad))` right?

Reply:
If you're using a DPDD-formatted Object Table then you won't have `coord_ra` and `coord_dec`,  you will just have `ra` and `dec`.

That's why there's the creation of the `ra` and `dec` columns above in `associate_object_ids_to_coadd` where we have a coadd ref catalog from the Butler (which is `coord_ra`, `coord_dec` in `u.rad`) and convert to `ra`, `dec`, in `u.deg` and then call `associate_object_ids_to_table`.